#  Python Language Basics

## FOLLOW PEP8!

In [1]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


## Language Semantics

Indentation, not braces:
Use **four spaces** as your default indentation and replacing tabs with four spaces.

In [2]:
"""for x in array:
    if x < pivot:
        less.append(x)
    else:
        greater.append(x)"""

'for x in array:\n    if x < pivot:\n        less.append(x)\n    else:\n        greater.append(x)'

Python statements also do not need to be terminated by semicolons. 

Semicolons can be used, however, to separate multiple statements on a single line (Putting multiple statements on one line is generally discouraged in Python):

In [3]:
a = 5
b = 6
c = 7

#### Multi-Line Statements

In [4]:
one = 2
two = 3
three = 5

In [5]:
total = one + two + three

In [6]:
total

10

## Comments

Any text preceded by the hash mark (pound sign) **#** is ignored by the Python interpreter.

In [7]:
print("Reached this line")  # Simple status report

Reached this line


## Variables and argument passing

In [8]:
a = [1, 2, 3]

When assigning a variable (or name) in Python, you are creating a __reference__ to the
object on the righthand side of the equals sign.

In [9]:
b = a

In [10]:
a.append(4)

In [11]:
b

[1, 2, 3, 4]

![alt text](images/2-1.png "Two references for the same object")

When you pass objects as arguments to a function, new local variables are created referencing the original objects without any copying. If you bind a new object to a variable inside a function, that change will not be reflected in the parent scope:

In [12]:
def append_element(some_list, element):
    some_list.append(element)

In [13]:
data = [1, 2, 3]

In [14]:
append_element(data, 4)

In [15]:
data

[1, 2, 3, 4]

## Dynamic references, strong types

Object references in Python have no type associated with them:

In [16]:
a = 5
type(a)

int

In [17]:
a = "foo"
type(a)

str

Python is considered a strongly typed language:

In [18]:
"5" + 5

TypeError: can only concatenate str (not "int") to str

Implicit conversions will occur only in certain obvious circumstances, such as the following:

In [19]:
a = 4.5
b = 2

In [20]:
print("a is {0}, b is {1}".format(type(a), type(b)))

a is <class 'float'>, b is <class 'int'>


In [21]:
a / b

2.25

You can check that an object is an instance of a particular type using the **isinstance** function:

In [22]:
isinstance(a, int)

False

**isinstance** can accept a tuple of types if you want to check that an object’s type is
among those present in the tuple:

In [23]:
a = 5
b = 4.5

In [24]:
isinstance(a, (int, float))

True

Assignments can be done on more than one variable "simultaneously" on the same line like this:

In [25]:
a, b = 3, 4

Swap can be done like this:

In [26]:
a, b = 1, 2

In [27]:
b, a = a, b

In [28]:
a, b

(2, 1)

## Binary operators and comparisons

In [29]:
5 - 7

-2

In [30]:
12 + 21.5

33.5

In [31]:
5 <= 2

False

To check if two references refer to the same object, use the **is** / **is not** keyword. 

In [32]:
a = [1, 2, 3]
b = a
c = list(a)

In [33]:
a is b

True

In [34]:
a is not c

True

Comparing with is is not the same as the **==** operator:

In [35]:
a == c

True

A very common use of __is__ and __is not__ is to check if a variable is __None__:

In [36]:
a = None

In [37]:
a is None

True

![alt text](images/2-2.png "Binary operators")

## Scalar Types (built-in data types)

Python along with its standard library has a small set of built-in types for handling numerical data, strings, boolean (True or False) values, and dates and time. These “single value” types are sometimes called ___scalar types___.

![alt text](images/2-3.png "Standard Python scalar types")

### Numeric types

The primary Python types for numbers are ___int___ and __*float*__:

In [38]:
ival = 17239871

In [39]:
ival**6

26254519291092456596965462913230729701102721

In [40]:
fval = 7.243

In [41]:
fval2 = 6.78e-5

Integer division not resulting in a whole number will always yield a floating-point number:

In [42]:
3 / 2

1.5

To get C-style integer division (which drops the fractional part if the result is not a whole number), use the floor division operator **//**:

In [43]:
3 // 2

1

Another operator available is the modulo (%) operator, which returns the integer remainder of the division. dividend % divisor = remainder.

In [44]:
11 % 3

2

You can also delete the reference to a number object by using the **del** statement.

In [45]:
var1 = 1
var2 = 10

In [46]:
del var1, var2

## Strings

You can write string literals using either single quotes ' or double quotes ":

In [47]:
a = "one way of writing a string"
b = "another way"

For multiline strings with line breaks, you can use triple quotes, either **' ' '** or **" " "**:

In [48]:
c = """
This is a longer string that
spans multiple lines
"""

The line breaks after " " " and after lines are included in the string:

In [49]:
c.count("\n")

3

Python strings are immutable; you cannot modify a string:

In [50]:
a = "this is a string"
a[10] = "f"

TypeError: 'str' object does not support item assignment

In [51]:
b = a.replace("string", "longer string")

Afer this operation, the variable a is unmodified:

In [52]:
a, b

('this is a string', 'this is a longer string')

Many Python objects can be converted to a string using the str function:

In [53]:
a = 5.6
s = str(a)

In [54]:
print(s)

5.6


Strings are a sequence of Unicode characters and therefore can be treated like other
sequences, such as lists and tuples:

In [55]:
s = "python"

In [56]:
list(s)

['p', 'y', 't', 'h', 'o', 'n']

In [57]:
s[:3]

'pyt'

The backslash character **\** is an ___escape___ character, meaning that it is used to specify special characters like newline \n or Unicode characters. To write a string literal with backslashes, you need to escape them:

In [58]:
s = "12\34"

In [59]:
s

'12\x1c'

In [60]:
s = "12\\34"

In [61]:
s

'12\\34'

If you have a string with a lot of backslashes and no special characters, you might find
this a bit annoying. Fortunately you can preface the leading quote of the string with **r**,
which means that the characters should be interpreted as is:

In [62]:
s = r"this\has\no\special\characters"

In [63]:
s

'this\\has\\no\\special\\characters'

Adding two strings together concatenates them and produces a new string:

In [64]:
a = "this is the first half "
b = "and this is the second half"

In [65]:
a + b

'this is the first half and this is the second half'

In [66]:
str = "Hello World!"

print(str)  # Prints complete string
print(str[0])  # Prints first character of the string
print(str[2:5])  # Prints characters starting from 3rd to 5th
print(str[2:])  # Prints string starting from 3rd character
print(str * 2)  # Prints string two times
print(str + "TEST")  # Prints concatenated string

Hello World!
H
llo
llo World!
Hello World!Hello World!
Hello World!TEST


## String formatting

In [67]:
template = "{0:.2f} {1:s} are worth US${2:d}"

In [68]:
template.format(4.5560, "Argentine Pesos", 1)

'4.56 Argentine Pesos are worth US$1'

* {0:.2f} means to format the first argument as a floating-point number with two decimal places.
* {1:s} means to format the second argument as a string.
* {2:d} means to format the third argument as an exact integer.

In [69]:
name = "John"
print("Hello, %s!" % name)

Hello, John!


In [70]:
name = "John"
age = 23
print("%s is %d years old." % (name, age))

John is 23 years old.


In [71]:
mylist = [1, 2, 3]
print("A list: %s" % mylist)

A list: [1, 2, 3]


* %s - String (or any object with a string representation, like numbers)

* %d - Integers

* %f - Floating point numbers

* % .\\<number of digits\\> f - Floating point numbers with a fixed amount of digits to the right of the dot.

* %x/%X - Integers in hex representation (lowercase/uppercase)

## New f-string formatting

In [72]:
temperature = 34.1234

In [73]:
print(f"Temprature is {temperature:.2f}")

Temprature is 34.12


## Bytes and Unicode

In modern Python (i.e., Python 3.0 and up), Unicode has become the first-class string
type to enable more consistent handling of ASCII and non-ASCII text. In older ver‐
sions of Python, strings were all bytes without any explicit Unicode encoding. You
could convert to Unicode assuming you knew the character encoding:

In [74]:
val = "español"

In [75]:
val

'español'

We can convert this Unicode string to its UTF-8 bytes representation using the encode method:

In [76]:
val_utf8 = val.encode("utf-8")

In [77]:
val_utf8

b'espa\xc3\xb1ol'

In [78]:
type(val_utf8)

bytes

Assuming you know the Unicode encoding of a bytes object, you can go back using
the decode method:

In [79]:
val_utf8.decode("utf-8")

'español'

While it’s become preferred to use UTF-8 for any encoding, for historical reasons you
may encounter data in any number of different encodings:

In [80]:
val.encode("latin1")

b'espa\xf1ol'

In [81]:
val.encode("utf-16")

b'\xff\xfee\x00s\x00p\x00a\x00\xf1\x00o\x00l\x00'

In [82]:
val.encode("utf-16le")

b'e\x00s\x00p\x00a\x00\xf1\x00o\x00l\x00'

It is most common to encounter bytes objects in the context of working with files, where implicitly decoding all data to Unicode strings may not be desired.
Though you may seldom need to do so, you can define your own byte literals by prefixing a string with **b**:

In [83]:
bytes_val = b"this is bytes"
bytes_val

b'this is bytes'

In [84]:
decoded = bytes_val.decode("utf8")
decoded  # this is str (Unicode) now

'this is bytes'

## Booleans

In [85]:
True and True

True

In [86]:
False or True

True

## Type casting

The **str**, **bool**, **int**, and **float** types are also functions that can be used to cast values to those types:

In [87]:
s = "3.14159"
fval = float(s)

In [88]:
type(fval)

float

In [89]:
int(fval)

3

In [90]:
bool(fval)

True

In [91]:
bool(0)

False

## None

**None** is the Python null value type. If a function does not explicitly return a value, it
implicitly returns **None**:

In [92]:
a = None
a is None

True

In [93]:
b = 5
b is not None

True

**None** is also a common default value for function arguments:

In [94]:
def add_and_maybe_multiply(a, b, c=None):
    result = a + b
    if c is not None:
        result = result * c
    return result

**None** is not only a reserved keyword but also a unique instance of **NoneType**:

In [95]:
type(None)

NoneType

## Dates and times

The built-in Python **datetime** module provides **datetime**, **date**, and **time** types. The **datetime** type, combines the information stored in date and time and is the most commonly used:

In [96]:
from datetime import datetime, date, time

In [97]:
dt = datetime(2011, 10, 29, 20, 30, 21)

In [98]:
dt.day, dt.second, dt.minute

(29, 21, 30)

Given a **datetime** instance, you can extract the equivalent **date** and **time** objects by
calling methods on the **datetime** of the same name:

In [99]:
dt.date(), dt.time()

(datetime.date(2011, 10, 29), datetime.time(20, 30, 21))

The **strftime** method formats a datetime as a string:

In [100]:
dt.strftime("%m/%d/%Y %H:%M")

'10/29/2011 20:30'

Strings can be converted (parsed) into **datetime** objects with the **strptime** function:

In [101]:
datetime.strptime("20091031", "%Y%m%d")

datetime.datetime(2009, 10, 31, 0, 0)

![alt text](images/2-4.png "Datetime format specification (ISO C89 compatible)")

Replacing time fields of a series of datetimes:

In [102]:
dt.replace(minute=0, second=0)

datetime.datetime(2011, 10, 29, 20, 0)

The difference of two **datetime** objects produces a **datetime.timedelta** type:

In [103]:
dt2 = datetime(2011, 11, 15, 22, 30)

In [104]:
delta = dt2 - dt

In [105]:
delta, type(delta)

(datetime.timedelta(days=17, seconds=7179), datetime.timedelta)

The output **timedelta(17, 7179)** indicates that the timedelta encodes an offset of 17
days and 7,179 seconds.

Adding a **timedelta** to a **datetime** produces a new shifted datetime:

In [106]:
dt, dt + delta

(datetime.datetime(2011, 10, 29, 20, 30, 21),
 datetime.datetime(2011, 11, 15, 22, 30))

### Types of Operator

Arithmetic Operators

Comparison (Relational) Operators

Assignment Operators

Logical Operators

Bitwise Operators

Membership Operators

Identity Operators

#### Python Arithmetic Operators

![alt text](images/ArithmeticOperators.png "Arithmetic Operators")

#### Python Comparison Operators

![alt text](images/ComparisonOperators.png "Comparison Operators")

#### Python Assignment Operators

![alt text](images/AssignmentOperators.png "Assignment Operators")

#### Python Bitwise Operators

![alt text](images/BitwiseOperators.png "Bitwise Operators")

#### Python Logical Operators

![alt text](images/LogicalOperators.png "Logical Operators")

#### Python Membership Operators

![alt text](images/MembershipOperators.png "Membership Operators")

#### Python Identity Operators

![alt text](images/IdentityOperators.png "Identity Operators")